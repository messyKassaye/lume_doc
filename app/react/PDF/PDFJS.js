/* eslint-disable import/no-mutable-exports, global-require, prefer-destructuring */
import { isClient } from 'app/utils';

let PDFJS = {};
let pdfjsLib = {};
let textLayerFactory = {};
if (isClient) {
  require('../../../node_modules/pdfjs-dist/web/pdf_viewer.css');

  PDFJS = require('../../../node_modules/pdfjs-dist/web/pdf_viewer.js');
  if (process.env.HOT || process.env.NODE_ENV === 'test') {
    pdfjsLib = require('pdfjs-dist');
    // pdfjsLib.workerSrc = 'http://localhost:8080/pdf.worker.js';
  } else {
    pdfjsLib = require('pdfjs-dist/webpack');
  }
  textLayerFactory = new PDFJS.DefaultTextLayerFactory();
}

export default { ...PDFJS, ...pdfjsLib };

export { textLayerFactory };
