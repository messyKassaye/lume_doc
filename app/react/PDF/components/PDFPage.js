import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { isClient } from 'app/utils';
import { PageReferences } from 'app/Viewer/components/PageReferences';
import PDFJS, { textLayerFactory } from '../PDFJS';

class PDFPage extends Component {
  constructor(props) {
    super(props);
    this.state = { rendered: false };
  }

  componentDidMount() {
    this.scrollCallback = this.scroll.bind(this);

    if (this.pageContainer && this.pageShouldRender()) {
      this.renderPage();
    }

    this.props.getViewportContainer().addEventListener('scroll', this.scrollCallback);
    this._mounted = true;
  }

  componentWillUnmount() {
    if (this.pdfPageView) {
      this.pdfPageView.destroy();
    }
    this.props.getViewportContainer().removeEventListener('scroll', this.scrollCallback);
    this._mounted = false;
  }

  scroll() {
    if (this.pageShouldRender() && !this.state.rendered) {
      this.renderPage();
    }

    if (!this.pageShouldRender() && this.pdfPageView) {
      this.pdfPageView.cancelRendering();
      this.pdfPageView.destroy();
      if (this.state.rendered) {
        this.props.onUnload(this.props.page);
      }
      this.setState({ rendered: false });
    }
  }

  pageShouldRender() {
    const pageRectangle = this.pageContainer.getBoundingClientRect();
    const vWidth = window.innerWidth || document.documentElement.clientWidth;
    const vHeight = window.innerHeight || document.documentElement.clientHeight;

    if (
      pageRectangle.right < 0 ||
      pageRectangle.bottom < -500 ||
      pageRectangle.left > vWidth ||
      pageRectangle.top > vHeight + 500
    ) {
      this.props.onHidden(this.props.page);
      return false;
    }
    this.checkVisibility(pageRectangle);
    return true;
  }

  checkVisibility(pageRectangle) {
    const viewportRect = this.props.getViewportContainer().getBoundingClientRect();

    const relativeElementRect = {
      top: pageRectangle.top - viewportRect.top,
      bottom: pageRectangle.bottom,
    };

    const offsetTop = relativeElementRect.top < 0 ? -relativeElementRect.top : 0;
    const offsetBottom =
      pageRectangle.bottom - viewportRect.bottom > 0
        ? pageRectangle.bottom - viewportRect.bottom
        : 0;
    const visibility = pageRectangle.height - offsetTop - offsetBottom;

    if (visibility > 0) {
      this.props.onVisible(this.props.page, visibility);
    }

    if (visibility < 0) {
      this.props.onHidden(this.props.page);
    }
  }

  renderReferences() {
    if (this.state.rendered) {
      return <PageReferences page={this.props.page} onClick={this.props.highlightReference} />;
    }

    return false;
  }

  renderPage() {
    if (!this.state.rendered && this.pdfPageView) {
      this.props.onLoading(this.props.page);
      this.pdfPageView.draw().catch(e => e);
      this.setState({ rendered: true });

      return;
    }
    if (!this.state.rendered) {
      this.props.onLoading(this.props.page);
      this.setState({ rendered: true });
      this.props.pdf.getPage(this.props.page).then(page => {
        const scale = 1;

        this.pdfPageView = new PDFJS.PDFPageView({
          container: this.pageContainer,
          id: this.props.page,
          scale,
          defaultViewport: page.getViewport({ scale }),
          enhanceTextSelection: true,
          textLayerFactory,
        });

        this.pdfPageView.setPdfPage(page);
        this.pdfPageView
          .draw()
          .then(() => {
            if (this._mounted) {
              this.setState({ height: this.pdfPageView.viewport.height });
            }
          })
          .catch(e => e);
      });
    }
  }

  render() {
    const style = { height: 1100 };
    if (this.state && this.state.height) {
      style.height = this.state.height + 20;
    }
    return (
      <div
        id={`page-${this.props.page}`}
        className="doc-page"
        ref={ref => {
          this.pageContainer = ref;
        }}
        style={style}
      >
        {this.renderReferences()}
      </div>
    );
  }
}

PDFPage.defaultProps = {
  getViewportContainer: () => (isClient ? document.querySelector('.document-viewer') : null),
  onVisible: () => {},
  onHidden: () => {},
  highlightReference: () => {},
};

PDFPage.propTypes = {
  getViewportContainer: PropTypes.func,
  page: PropTypes.number.isRequired,
  onVisible: PropTypes.func,
  onHidden: PropTypes.func,
  onLoading: PropTypes.func.isRequired,
  onUnload: PropTypes.func.isRequired,
  pdf: PropTypes.object.isRequired,
  highlightReference: PropTypes.func,
};

export default PDFPage;
