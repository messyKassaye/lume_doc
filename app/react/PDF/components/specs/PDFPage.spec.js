/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';

import PDFJS from '../../PDFJS';
import PDFPage from '../PDFPage';

describe('PDFPage', () => {
  let component;
  let instance;
  let container;
  const pdfObject = { numPages: 2 };

  let props;

  beforeEach(() => {
    container = document.createElement('div');
    container.className = 'document-viewer';
    spyOn(PDFJS, 'getDocument').and.returnValue(Promise.resolve(pdfObject));
    props = {
      onLoading: jasmine.createSpy('onLoading'),
      onUnload: jasmine.createSpy('onUnload'),
      page: 2,
      getViewportContainer: () => container,
      pdf: {
        getPage: jest.fn().mockReturnValueOnce(Promise.resolve({ getViewport: jest.fn() })),
      },
    };
  });

  const render = () => {
    component = shallow(<PDFPage {...props} />);
    instance = component.instance();
  };

  describe('render', () => {
    it('should render a div as pageContainer', () => {
      render();
      expect(component).toMatchSnapshot();
    });

    it('should pass pass proper height when state.height', () => {
      render();
      expect(component.setState({ height: 5 })).toMatchSnapshot();
    });
  });

  describe('pageShouldRender', () => {
    it('should return true by default', () => {
      render();
      instance.pageContainer = { getBoundingClientRect: () => ({}) };
      expect(instance.pageShouldRender()).toBe(true);
    });

    it('should call onHidden when not being rendered', () => {
      props.onHidden = jasmine.createSpy('onHidden');
      render();
      instance.pageContainer = { getBoundingClientRect: () => ({ right: -1 }) };
      instance.pageShouldRender();
      expect(props.onHidden).toHaveBeenCalledWith(2);
    });

    it('should return false when page is not in viewport with a 500px offset', () => {
      render();
      instance.pageContainer = { getBoundingClientRect: () => ({ right: -1 }) };
      expect(instance.pageShouldRender()).toBe(false);

      instance.pageContainer = { getBoundingClientRect: () => ({ bottom: -501 }) };
      expect(instance.pageShouldRender()).toBe(false);

      window.innerWidth = 5;
      instance.pageContainer = { getBoundingClientRect: () => ({ left: 6 }) };
      expect(instance.pageShouldRender()).toBe(false);

      window.innerHeight = 5;
      instance.pageContainer = { getBoundingClientRect: () => ({ top: 506 }) };
      expect(instance.pageShouldRender()).toBe(false);
    });
  });

  describe('renderPage', () => {
    describe('when its not rendered but pdfPageView exists', () => {
      it('should draw the page and call onLoading', () => {
        render();
        instance.rendered = false;
        instance.pdfPageView = {
          draw: jasmine.createSpy('draw').and.returnValue(Promise.resolve()),
        };
        instance.renderPage();
        expect(instance.pdfPageView.draw).toHaveBeenCalled();
        expect(instance.state.rendered).toBe(true);
        expect(props.onLoading).toHaveBeenCalledWith(props.page);
      });
    });

    describe('when its not rendered and no pdfPageView exists', () => {
      it('should create pdfPageView object and render the page', done => {
        render();
        instance.state.rendered = false;
        const pdfPageViewPrototype = {
          setPdfPage: jest.fn(),
          draw: jest.fn().mockReturnValueOnce(Promise.resolve()),
        };

        PDFJS.PDFPageView = function pdfPageView() {};
        PDFJS.DefaultTextLayerFactory = function pdfPageView() {};
        PDFJS.PDFPageView.prototype = pdfPageViewPrototype;

        instance.renderPage();

        setTimeout(() => {
          expect(props.onLoading).toHaveBeenCalledWith(props.page);
          expect(instance.state.rendered).toBe(true);
          expect(pdfPageViewPrototype.setPdfPage).toHaveBeenCalled();
          expect(pdfPageViewPrototype.draw).toHaveBeenCalled();
          done();
        });
      });
    });
  });

  describe('scroll', () => {
    describe('when pageShouldRender', () => {
      it('should render page', () => {
        render();
        spyOn(instance, 'pageShouldRender').and.returnValue(true);
        spyOn(instance, 'renderPage');
        instance.scroll();
        expect(instance.renderPage).toHaveBeenCalled();
      });

      it('should calculate page visibility and if visible call onVisible callback', () => {
        props.onVisible = jasmine.createSpy('onVisible');

        render();
        component.setProps({
          getViewportContainer: () => ({
            getBoundingClientRect: () => ({ top: 100, height: 500, bottom: 600 }),
          }),
        });

        instance.pageContainer = {
          getBoundingClientRect: () => ({ top: 100, height: 1000, bottom: 1100 }),
        };
        instance.pageShouldRender();

        let expectedPage = 2;
        let pageVisibility = 500;
        expect(props.onVisible).toHaveBeenCalledWith(expectedPage, pageVisibility);

        instance.pageContainer = {
          getBoundingClientRect: () => ({ top: -600, height: 1000, bottom: 400 }),
        };
        component.update();
        instance.pageShouldRender();

        expectedPage = 2;
        pageVisibility = 300;
        expect(props.onVisible).toHaveBeenCalledWith(expectedPage, pageVisibility);
      });

      it('should call pageHidden when there is no visibility', () => {
        props.onVisible = jasmine.createSpy('onVisible');
        props.onHidden = jasmine.createSpy('onHidden');

        render();
        component.setProps({
          viewportContainer: {
            getBoundingClientRect: () => ({ top: 100, height: 500, bottom: 600 }),
          },
        });

        instance.pageContainer = {
          getBoundingClientRect: () => ({ top: -2100, height: 1000, bottom: 1100 }),
        };
        instance.pageShouldRender();

        expect(props.onVisible).not.toHaveBeenCalled();
        expect(props.onHidden).toHaveBeenCalledWith(2);
      });
    });

    describe('when not pageShouldRender', () => {
      it('should check if its rendered and unload and set rendered property to false', () => {
        render();
        instance.pdfPageView = jasmine.createSpyObj(['cancelRendering', 'destroy']);
        spyOn(instance, 'pageShouldRender').and.returnValue(false);
        instance.scroll();
        expect(instance.pdfPageView.cancelRendering).toHaveBeenCalled();
        expect(instance.pdfPageView.destroy).toHaveBeenCalled();
        expect(instance.state.rendered).toBe(false);
      });

      describe('when its rendered', () => {
        it('should call onUnload with the pageNumber and set rendered property to false', () => {
          render();
          instance.pdfPageView = jasmine.createSpyObj(['cancelRendering', 'destroy']);
          instance.state.rendered = true;
          spyOn(instance, 'pageShouldRender').and.returnValue(false);
          instance.scroll();
          expect(props.onUnload).toHaveBeenCalledWith(props.page);
          expect(instance.state.rendered).toBe(false);
        });
      });
    });
  });
});
