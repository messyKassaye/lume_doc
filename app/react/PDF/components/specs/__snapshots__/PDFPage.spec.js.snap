// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFPage render should pass pass proper height when state.height 1`] = `
<div
  className="doc-page"
  id="page-2"
  style={
    Object {
      "height": 25,
    }
  }
/>
`;

exports[`PDFPage render should render a div as pageContainer 1`] = `
<div
  className="doc-page"
  id="page-2"
  style={
    Object {
      "height": 1100,
    }
  }
/>
`;
